feat: add first rate limiting article (#594)
fiberplane/fiberplane: www/src/content/blog/2025-05-13-An-introduction-to-rate-limiting.mdx

- Adding popular third-party rate limiters to a Hono app
- Rate limiting algorithms, with real-world Redis implementations
- Rolling your own rate limiting solutions

By the end of this series, you will be able to confidently pick the right limiter(s) for your use-case, or even roll your own! First though, let’s review the whats, whys, and hows of rate limiting.

## What is rate limiting anyway?

In practice, rate limiting refers to a broad variety of policies and implementations with equally-diverse goals. In principle though, these are united by a set of common factors. In brief, rate limiters control how often a resource can be accessed, and by whom. 

To accomplish this, most limiters use some kind of identifier (e.g., user ID or client IP) to track requests from a given source, and an algorithm to determine if and when requests from each source should be allowed.

The specifics vary by algorithm and implementation, but most limiters are defined by:

- A **limit** on how many requests can be made within a defined **window** of time
- The **cost** of each request counted against the limit
- How often a client’s limit is **refilled** or **reset**, allowing additional requests

> Client consumption is measured in **tokens (**or as the sum of request timestamps), often stored in an in-memory database like [Redis](https://redis.io/) or [Memcached](https://memcached.org/). On each request, the rate limiting algorithm is used to determine whether the number of recent requests exceeds the limit.

Reviewer: ```suggestion
Client consumption is measured in **tokens** (or as the sum of request timestamps), often stored in an in-memory database like [Redis](https://redis.io/) or [Memcached](https://memcached.org/). On each request, the rate limiting algorithm is used to determine whether the number of recent requests exceeds the limit.
```

> In-memory databases are used to reduce the latency introduced by a rate limiting layer, but limiters inevitably increase response times—albeit minimally. They play a crucial role in capacity management and application security though, preventing the app from crashing under a flood of traffic, and mitigating an app’s vulnerability to certain attack vectors.

Reviewer: just a little tightening of this paragraph, hope that's okay. fewer adverbs and prepositions

```suggestion
In-memory databases help minimize the latency added by the rate-limiting layer, but some overhead is unavoidable. Even so, rate limiters are essential for managing system capacity and securing applications—they help prevent crashes during traffic spikes and reduce exposure to certain types of attacks.
```

## Identifying Clients

In practice, apps often use a combination of limiters—each using a different identifier—to address these different vulnerabilities. IP-based limiters can protect pre-login workflows, for example, while location-based limits can help manage app capacity across regions.

To pick the right identifier for each use-case, it’s important to understand the strengths and limitations of each, as well as how these can support different requirements and strategies. As developers, we must also be mindful of the ethical and legal implications each identifier entails.

> ### Ethical Considerations

Reviewer: ```suggestion
### Ethical and Legal Considerations
```

seeing as this section also covers the legal aspects (and distinguishes them) makes sense to title it as aboth "Ethical and Legal" imo

> Storing personally-identifiable user data makes users vulnerable to doxxing, targeted cyber attacks, and government overreach. It also habituates users (and developers) to an ongoing erosion of digital privacy that threatens their safety and political freedom.

Reviewer: although i don't disagree, i feel like there's a tonal mismatch in having a sole clause about threatening safety and freedom

a few ideas for alts:

- change "threatens their" to "jeopardizes" 
- strike the "that threatens..." clause, but link to an article that goes more in-depth on social/ethical/legal risks inherent to storing more PII than you need

Author: can you say more? the clause is meant to be about the way developer decisions shape the kinds of digital interactions users become accustomed to. i.e., how we influence the world of tomorrow

Reviewer: to be clear the note is specifically about the clause "that threatens their safety and political freedom"

basically, it felt strange to be reading an article on rate limiting, and then finish a paragraph that punches really hard (with its rhetoric) on political impact. at one moment i'm thinking of storing PII, the next moment i'm thinking about my safety.

maybe as a point of contrast, saying it "makes users vulnerable to doxxing, targeted cyber attacks, and government overreach" is a concrete, easily attributable risk associated with being overeager with PII.

tl;dr, in my mind, threat-rhetoric about safety and freedom feels out of place in this paragraph, and would feel more in-character for a piece that deals with that as a central theme.

still open to discussion

Author: i'm still not sure i understand. is it that you find "threat" overly-emotive in this context? or is it more that "threatening safety and political freedom" isn't sufficiently concrete? or that the practical consequences of eroding digital privacy are complex and varied?

Author: i'm not married to the language, but i do think it's important to communicate that getting comfortable with trading privacy for convenience has dire consequences for individuals and the world at large. the internet is constantly evolving, and devs are ultimately responsible for whether that's for the better or worse.

Account or user IDs are the most broadly useful identifier, as they can be used to meet an app’s business needs, while also providing some protection against resource exhaustion and malicious request patterns. 

Note that while an ID uniquely identifies an *account*, it doesn’t inherently reveal anything about the user behind it or their location, making it a solid choice from an ethical perspective as well.

### Request IPs

Many examples use the client IP, as it’s relationship to requests is fairly intuitive, and the implementation doesn’t rely on the app’s business logic (e.g., for a user ID). IP limits are sub-optimal for capacity management though, as they do not necessarily correspond to unique users.

Multiple users could share the same IP, in which case one user could exhaust the limit for all. Conversely, a single client could also circumvent limits by accessing services using different IPs by using a VPN or a proxy.

> While IPs should not be used as a primary limiter, they play a vital role in security-oriented rate limiting, especially in auth flows. If a user hasn’t yet been authenticated (or even registered), IP limits can prevent a malicious client from bombarding auth services, either to overwhelm them or in an effort to brute-force access. In more sophisticated rate limiting implementations, IPs with a pattern of malicious requests can also be blacklisted.

Reviewer: ```suggestion
While IPs should not be used as a primary limiter, they play a vital role in security-oriented rate limiting, especially in auth flows. If a user hasn’t yet been authenticated (or even registered), IP limits can prevent a malicious client from bombarding auth services, either to overwhelm them or in an effort to brute-force access. In more sophisticated rate limiting implementations, IPs with a pattern of malicious requests can also be blocked.
```

### Geolocation

Location-based limiting is far less common outside of enterprise-scale apps, perhaps in part because it is comparatively resource intensive. It can be used to dynamically manage limits by region based on expected traffic patterns. If request count decreases at night, for example, the rate limit for a given region could be lowered until the next morning.

This type of limiting is especially useful in a security context, as it can mitigate (and flag) suspicious request patterns. Doing so effectively requires some kind of geo-fencing though, as well as a service dedicated to managing the dynamic limits.

### Resource Limiting

> Of course, it’s also possible to limit how much a resource is used at all, whether it’s a server, an  endpoint, or a single workflow. Since resource limiters don’t discriminate between clients, they’re not recommended for most rate limiting use-cases. They’re mostly useful for capacity management, and serve as a last resort against unexpected floods of traffic.

Reviewer: ```suggestion
Of course, it’s also possible to limit how much a resource is used at all, whether it’s a server, an endpoint, or a single workflow. Since resource limiters don’t discriminate between clients, they’re not recommended for most rate limiting use-cases. They’re mostly useful for capacity management, and serve as a last resort against unexpected floods of traffic.
```

The `RateLimit` header family is often used to provide additional detail, which can help clients avoid hitting the limit to begin with. This trades security for user experience, so it’s mostly used in responses to data requests or other business-limited operations.

- **`RateLimit-Limit`**: How many requests the client is allowed within a given window.
- **`RateLimit-Remaining`**: The number allowed of requests remaining (`limit - consumed`).
- **`RateLimit-Reset`**: Optionally, when the limit will be refilled or reset, in seconds or as a timestamp.

The best way to communicate rate limit data ultimately depends on the role the limiter is playing, as well as the app’s requirements. Third-party limiters may use different patterns or standards to communicate rate limit information, so it’s important to consider what needs to be shared (and with whom) when choosing a solution.

> ## From theory to praxis

Reviewer: ```suggestion
## From theory to practice
```

might opt to just use practice, since praxis is a more academic term

As you can see, rate limiting is a complex and diverse space. There’s a lot that I didn’t cover, but some of those blanks will be filled in when we take a closer look at rate limiting algorithms and implementations in subsequent articles.

> While not exhaustive, this summary of core rate limiting motivations and strategies is meant as a foundation for these deeper dives that follow. If you think I should have addressed something in greater depth—or if I missed it altogether—please let me know in the comments!

Reviewer: "let me know in the comments" won't make much sense on our blog 😅 — CTA could be to open an issue on the monorepo instead? or `@fiberplane` on twitter
